mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala

```diff
@@ -19,45 +19,133 @@ package org.apache.spark.ml.feature
+  protected def validateAndTransformSchema(schema: StructType): StructType = {
+    SchemaUtils.checkColumnType(schema, $(inputCol), new ArrayType(StringType, true))
+    SchemaUtils.appendColumn(schema, $(outputCol), new VectorUDT)
+  }
+
+  override def validateParams(): Unit = {
+    require($(vocabSize) > 0, s"The vocabulary size (${$(vocabSize)}) must be above 0.")
+  }
+}
 
 /**
  * :: Experimental ::
- * Converts a text document to a sparse vector of token counts.
- * @param vocabulary An Array over terms. Only the terms in the vocabulary will be counted.
+ * Extracts a vocabulary from document collections and generates a [[CountVectorizerModel]].
  */
-@Experimental
-class CountVectorizerModel (override val uid: String, val vocabulary: Array[String])
-  extends UnaryTransformer[Seq[String], Vector, CountVectorizerModel] {
+class CountVectorizer(override val uid: String)
+  extends Estimator[CountVectorizerModel] with CountVectorizerParams {
 
-  def this(vocabulary: Array[String]) =
-    this(Identifiable.randomUID("cntVec"), vocabulary)
+  def this() = this(Identifiable.randomUID("cntVec"))
 
   /**
-   * Corpus-specific filter to ignore scarce words in a document. For each document, terms with
-   * frequency (count) less than the given threshold are ignored.
+   * The minimum number of times a token must appear in the corpus to be included in the vocabulary
    * Default: 1
    * @group param
    */
-  val minTermFreq: IntParam = new IntParam(this, "minTermFreq",
-    "minimum frequency (count) filter used to neglect scarce words (>= 1). For each document, " +
-      "terms with frequency less than the given threshold are ignored.", ParamValidators.gtEq(1))
+  val minCount: IntParam = new IntParam(this, "minCount",
+    "minimum number of times a token must appear in the corpus to be included in the vocabulary."
+    , ParamValidators.gtEq(1))
+
+  /** @group getParam */
+  def getMinCount: Int = $(minCount)
 
   /** @group setParam */
-  def setMinTermFreq(value: Int): this.type = set(minTermFreq, value)
+  def setInputCol(value: String): this.type = set(inputCol, value)
 
-  /** @group getParam */
-  def getMinTermFreq: Int = $(minTermFreq)
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
 
-  setDefault(minTermFreq -> 1)
+  /** @group setParam */
+  def setVocabSize(value: Int): this.type = set(vocabSize, value)
+
+  /** @group setParam */
+  def setMinCount(value: Int): this.type = set(minCount, value)
 
-  override protected def createTransformFunc: Seq[String] => Vector = {
+  setDefault(vocabSize -> 10000, minCount -> 1)
+
+  override def fit(dataset: DataFrame): CountVectorizerModel = {
+    transformSchema(dataset.schema, logging = true)
+    val input = dataset.select($(inputCol)).map(_.getAs[Seq[String]](0))
+    val wordCounts: RDD[(String, Long)] = input
+      .flatMap { case (tokens) => tokens.map(_ -> 1L) }
+      .reduceByKey(_ + _)
+      .filter(_._2 >= $(minCount))
```

Comment:
This and other uses of Params should be made into local copies before RDD operations, like you had originally.  There is a minor cost to the lookup, and it's good practice to prevent the whole class from being serialized (though it does not matter for this case).